java/src/main/java/io/cucumber/gherkin/PickleCompiler.java

```diff
@@ -98,25 +98,34 @@ private void compileRule(List<Pickle> pickles, Rule rule, List<Tag> parentTags,
     }
 
     private void compileScenario(List<Pickle> pickles, Scenario scenario, List<Tag> parentTags, List<Step> backgroundSteps, String language, String uri) {
-        List<PickleStep> steps = new ArrayList<>();
+        int pickleStepCount = scenario.getSteps().size() + backgroundSteps.size();
+        List<PickleStep> steps = new ArrayList<>(pickleStepCount);
         if (!scenario.getSteps().isEmpty()) {
-            List<Step> allSteps = new ArrayList<>();
+            // usually, a scenario has at least one step, but that's not mandatory
+            List<Step> allSteps = new ArrayList<>(pickleStepCount);
```

Comment:
I think we can avoid this array creation by duplicating the `allSteps` for loop.